src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -1222,24 +1222,47 @@ bool LibraryCallKit::inline_vector_mem_masked_operation(bool is_store) {
 
   int mem_num_elem = using_byte_array ? num_elem * type2aelembytes(elem_bt) : num_elem;
   BasicType mem_elem_bt = using_byte_array ? T_BYTE : elem_bt;
-  bool use_predicate = arch_supports_vector(is_store ? Op_StoreVectorMasked : Op_LoadVectorMasked,
-                                            mem_num_elem, mem_elem_bt,
-                                            (VectorMaskUseType) (VecMaskUseLoad | VecMaskUsePred));
-  // Masked vector store operation needs the architecture predicate feature. We need to check
-  // whether the predicated vector operation is supported by backend.
-  if (is_store && !use_predicate) {
-    if (C->print_intrinsics()) {
-      tty->print_cr("  ** not supported: op=storeMasked vlen=%d etype=%s using_byte_array=%d",
-                    num_elem, type2name(elem_bt), using_byte_array ? 1 : 0);
+  bool supports_predicate = arch_supports_vector(is_store ? Op_StoreVectorMasked : Op_LoadVectorMasked,
+                                                mem_num_elem, mem_elem_bt,
+                                                (VectorMaskUseType) (VecMaskUseLoad | VecMaskUsePred));
+
+  // If current arch does not support the predicated operations, we have to bail
+  // out when current case uses the predicate feature.
+  if (!supports_predicate) {
+    bool use_predicate = false;
+    if (is_store) {
+      // Masked vector store always uses the predicated store.
```

Comment:
Can masked store be implemented as Load + Blend + Store?

Reply:
The blend should be with the intended-to-store vector, so that masked lanes contain the need-to-store elements and unmasked lanes contain the loaded elements, which would be stored back, which results in unchanged values.

Reply:
Yeah, maybe I misunderstood what you mean. So maybe the masked store `(store(src, m))` could be implemented with:
```
1) v1 = load
2) v2 = blend(load, src, m)
3) store(v2)
```
Let's record this a JBS and fix it with a followed-up patch. Thanks!

Reply:
The optimization for masked store is recorded to: https://bugs.openjdk.java.net/browse/JDK-8284050

Reply:
> The blend should be with the intended-to-store vector, so that masked lanes contain the need-to-store elements and unmasked lanes contain the loaded elements, which would be stored back, which results in unchanged values.

It may not work if memory is beyond legal accessible address space of the process, a corner case could be a page boundary.  Thus re-composing the intermediated vector which partially contains actual updates but effectively perform full vector write to destination address may not work in some cases.

Reply:
Thanks for the comment! So how about adding the check for the valid array range like the masked vector load?
Codes like:
```
public final
    void intoArray(byte[] a, int offset,
                   VectorMask<Byte> m) {
        if (m.allTrue()) {
            intoArray(a, offset);
        } else {
            ByteSpecies vsp = vspecies();
            if (offset >= 0 && offset <= (a.length - vsp.length())) {     // a full range check
                intoArray0(a, offset, m, /* usePred */ false);                   // can be vectorized by load+blend_store
            } else {
                checkMaskFromIndexSize(offset, vsp, m, 1, a.length);
                intoArray0(a, offset, m, /* usePred */ true);                    // only be vectorized by the predicated store
            }
        }
    }
```

Reply:
Thanks, this looks ok since last condition will not be intrinsified if targets does not support predicated vector store.

Reply:
Hi @merykitty @jatin-bhateja , could you please help to take a review at the similar store masked PR https://github.com/openjdk/jdk/pull/8544 ? Any feedback is welcome! Thanks so much!